server/webhooks/failing_policies.go

```diff
@@ -17,101 +17,203 @@ import (
+				level.Error(logger).Log("msg", "failed to remove policy from set", "policyID", policyID, "err", err)
+			}
+			continue
+		case err != nil:
+			return ctxerr.Wrapf(ctx, err, "get policy: %d", policyID)
+		default:
+			// Ok
+		}
+
+		if policy.TeamID != nil {
+			// team policy
+			err := getTeam(*policy.TeamID)
+			switch {
+			case errors.Is(err, sql.ErrNoRows):
+				// shouldn't happen, unless the team was deleted after the policy was retrieved above
+				level.Debug(logger).Log("msg", "team does not exist", "teamID", *policy.TeamID)
+				continue
+			case err != nil:
+				level.Error(logger).Log("msg", "failed to get team", "teamID", *policy.TeamID, "err", err)
+				continue
+			}
+
+			settings := teamSettings[*policy.TeamID]
+			if !settings.Enable {
+				continue
+			}
+
+			_, ok := teamPolicyIDs[*policy.TeamID][policy.ID]
+			if !ok {
+				level.Debug(logger).Log("msg", "skipping failing policy, not found in team policy IDs", "policyID", policyID)
+				if err := failingPoliciesSet.RemoveSet(policy.ID); err != nil {
+					level.Error(logger).Log("msg", "failed to remove policy from set", "policyID", policyID, "err", err)
+				}
+				continue
+			}
+
+			webhookURL := teamWebhookURLs[*policy.TeamID]
+
+			err = sendFailingPoliciesBatchedPOSTs(
+				ctx,
+				policy,
+				failingPoliciesSet,
+				settings.HostBatchSize,
+				serverURL,
+				webhookURL,
+				now,
+				logger,
+			)
+			if err != nil {
+				level.Error(logger).Log("msg", "failed to send failing policies webhook requests", "policyID", policy.ID, "err", err)
+			}
+
+			continue
+		}
+
+		// global policy
+		err = sendFailingPoliciesBatchedPOSTs(
+			ctx,
+			policy,
+			failingPoliciesSet,
+			globalSettings.HostBatchSize,
+			serverURL,
+			globalWebhookURL,
+			now,
+			logger,
+		)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to send failing policies webhook requests", "policyID", policy.ID, "err", err)
 		}
 	}
-	return nil
-}
 
-type postData struct {
-	serverURL  *url.URL
-	now        time.Time
-	webhookURL string
+	return nil
 }
 
 func sendFailingPoliciesBatchedPOSTs(
 	ctx context.Context,
 	policy *fleet.Policy,
 	failingPoliciesSet fleet.FailingPolicySet,
-	postData postData,
 	hostBatchSize int,
+	serverURL *url.URL,
+	webhookURL *url.URL,
+	now time.Time,
 	logger kitlog.Logger,
 ) error {
 	hosts, err := failingPoliciesSet.ListHosts(policy.ID)
 	if err != nil {
-		return ctxerr.Wrapf(ctx, err, "listing hosts for global failing policies set %d", policy.ID)
+		return ctxerr.Wrapf(ctx, err, "listing hosts for failing policies set %d", policy.ID)
 	}
 	if len(hosts) == 0 {
-		level.Debug(logger).Log("id", policy.ID, "msg", "no hosts")
+		level.Debug(logger).Log("msg", "no hosts", "policyID", policy.ID)
 		return nil
 	}
-	if hostBatchSize == 0 {
-		hostBatchSize = len(hosts)
-	}
 	sort.Slice(hosts, func(i, j int) bool {
 		return hosts[i].ID < hosts[j].ID
 	})
-	for len(hosts) > 0 {
```

Comment:
Was there any bug with the previous batching implementation?

Reply:
There weren't any bugs, but I found it harder to read. In particular, with the previous method, the code for handling the batch was a bit more spread out

```
	for len(hosts) > 0 {
		j := hostBatchSize
		if l := len(hosts); j > l {
			j = l
		}
		batch := hosts[:j]
		// ...
		hosts = hosts[j:] 

```

This may be more of a stylistic choice, but I felt a more traditional for loop with initialization; condition; increment read a bit better. No objection to reverting back if you feel strongly about it.

Reply:
No worries, good thing we have tests (`TestSendBatchedPOSTs`).